docs/cloud-integration.md

```diff
@@ -0,0 +1,512 @@
+
+
+For use outside of Google cloud, `gcs-connector.jar` must be be manually downloaded then added
+to `$SPARK_HOME/jars`.
+
+
+## <a name="cloud_stores_are_not_filesystems"></a>Important: Cloud Object Stores are Not Real Filesystems
+
+Object stores are not filesystems: they are not a hierarchical tree of directories and files.
+
+The Hadoop filesystem APIs offer a filesystem API to the object stores, but underneath
+they are still object stores, [and the difference is significant](http://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/filesystem/introduction.html)
+
+While object stores can be used as the source and store
+for persistent data, they cannot be used as a direct replacement for a cluster-wide filesystem such as HDFS.
+This is important to know, as the fact they are accessed with the same APIs can be misleading.
+
+### Directory Operations May be Slow and Non-atomic
+
+Directory rename and delete may be performed as a series of operations. Specifically, recursive
+directory deletion may be implemented as "list the objects, delete them singly or in batches".
+File and directory renames may be implemented as "copy all the objects" followed by the delete operation.
+
+1. The time to delete a directory depends on the number of files in the directory.
+1. Directory deletion may fail partway through, leaving a partially deleted directory.
+1. Directory renaming may fail part way through, leaving the destination directory containing some of the files
+being renamed, the source directory untouched.
+1. The time to rename files and directories increases with the amount of data to rename.
+1. If the rename is done on the client, the time to rename
+each file will depend upon the bandwidth between client and the filesystem. The further away the client
+is, the longer the rename will take.
+1. Recursive directory listing can be very slow. This can slow down some parts of job submission
+and execution.
+
+Because of these behaviours, committing of work by renaming directories is neither efficient nor
+reliable. In Spark 1.6 and predecessors, there was a special output committer for Parquet,
+the `org.apache.spark.sql.execution.datasources.parquet.DirectParquetOutputCommitter`
+which bypasses the rename phase. However, as well as having major problems when used
+ with speculative execution enabled, it handled failures badly. For this reason, it
+[was removed from Spark 2.0](https://issues.apache.org/jira/browse/SPARK-10063).
+
+*Critical* speculative execution does not work against object
+stores which do not support atomic directory renames. Your output may get
+corrupted.
+
+*Warning* even non-speculative execution is at risk of leaving the output of a job in an inconsistent
+state if a "Direct" output committer is used and executors fail.
+
+### Data is Not Written Until the OutputStream's `close()` Operation.
+
+Data written to the object store is often buffered to a local file or stored in memory,
```

Comment:
(FWIW, where this does cause problems is that really slow writes can block things like heartbeating protocols if the same thread is doing the heartbeat. Hasn't surfaced in spark AFAIK.